sql/idempotent/010-vectorizer-int.sql

```diff
@@ -0,0 +1,409 @@
+-- _vectorizer_create_source_trigger
+create or replace function ai._vectorizer_create_source_trigger
+( trigger_name name
+, queue_schema name
+, queue_table name
+, source_schema name
+, source_table name
+, source_pk jsonb
+) returns void as
+$func$
+declare
+    _sql text;
+begin
+    select pg_catalog.format
+    ( $sql$
+    create function %I.%I() returns trigger
+    as $plpgsql$
+    begin
+        insert into %I.%I (%s)
+        values (%s);
+        return null;
+    end;
+    $plpgsql$ language plpgsql volatile parallel safe security invoker
+    set search_path to pg_catalog, pg_temp
+    $sql$
+    , queue_schema, trigger_name
+    , queue_schema, queue_table
+    , (
+        select pg_catalog.string_agg(pg_catalog.format('%I', x.attname), ', ' order by x.attnum)
+        from pg_catalog.jsonb_to_recordset(source_pk) x(attnum int, attname name)
+      )
+    , (
+        select pg_catalog.string_agg(pg_catalog.format('new.%I', x.attname), ', ' order by x.attnum)
+        from pg_catalog.jsonb_to_recordset(source_pk) x(attnum int, attname name)
+      )
+    ) into strict _sql
+    ;
+    execute _sql;
+
+    select pg_catalog.format
+    ( $sql$
+    create trigger %I
+    after insert or update
+    on %I.%I
+    for each row execute function %I.%I();
+    $sql$
+    , trigger_name
+    , source_schema, source_table
+    , queue_schema, trigger_name
+    ) into strict _sql
+    ;
+    execute _sql;
+end;
+$func$
+language plpgsql volatile security invoker
+set search_path to pg_catalog, pg_temp
+;
+
+-------------------------------------------------------------------------------
+-- _vectorizer_async_ext_job
+create or replace procedure ai._vectorizer_async_ext_job
+( job_id int default null
+, config jsonb default null
+) as
+$func$
+declare
+    _vectorizer_id int;
+    _queue_schema name;
+    _queue_table name;
+    _sql text;
+    _item bigint;
+begin
+    set local search_path = pg_catalog, pg_temp;
+    if config is null then
+        raise exception 'config is null';
+    end if;
+
+    -- get the vectorizer id from the config
+    select pg_catalog.jsonb_extract_path_text(config, 'vectorizer_id')::int
+    into strict _vectorizer_id
+    ;
+    -- look up the queue table for the vectorizer
+    select v.queue_schema, v.queue_table into strict _queue_schema, _queue_table
+    from ai.vectorizer v
+    where v.id operator(pg_catalog.=) _vectorizer_id
+    ;
+    -- if there is at least one item in the queue, we need to execute the vectorizer
+    select pg_catalog.format
+    ( $sql$
+    select 1
+    from %I.%I
+    for update skip locked
+    limit 1
+    $sql$
+    , _queue_schema, _queue_table
+    ) into strict _sql
+    ;
+    execute _sql into _item;
+    commit;
+    set local search_path = pg_catalog, pg_temp;
+    if _item is null then
+        -- nothing to do
+        return;
+    end if;
+    -- execute the vectorizer
+    perform ai.execute_vectorizer(_vectorizer_id);
```

Comment:
Should the function that's called here be configurable? Just thinking about how would it look like if I'm running the open source version and I want to use the TS job to schedule my own custom external vectorizer.

Reply:
yeah, i think we need something like that. not sure exactly how to do it just yet

Reply:
will remove in a separate PR. at least this forced me to implement in a way that supports multiple schedulers